core/src/main/java/org/elasticsearch/node/Node.java

```diff
@@ -310,6 +312,14 @@ public Node start() {
         // Start the transport service now so the publish address will be added to the local disco node in ClusterService
         TransportService transportService = injector.getInstance(TransportService.class);
         transportService.start();
+        DiscoveryNode localNode = injector.getInstance(DiscoveryNodeService.class)
+                .buildLocalNode(transportService.boundAddress().publishAddress());
+        clusterService.setLocalNode(localNode);
```

Comment:
I guess this trades off one kind of ugliness (having to remember to set the local node) with another  - i.e, set things here (I would remove the setLocalNode method) and have the start method be disabled. I think long term, with should figure something out for this problem. As we remove more of Guice will need this more often...